rfc/instance.md

```diff
@@ -0,0 +1,145 @@
+
+# Motivation
+
+The API changes allow better customization of low-level behavior. In turn, this will:
+
+1. Enables certain performance optimizations
+2. Makes it easier to integrate Spaniel with a wider range of environments
+
+This RFC is focused on the API changes themselves, not the optimizations, which will be added in the future.
+
+## Environment configuration
+
+We want to make it easy to configure spaniel for use in any environment:
+
+1. Test environments
+2. Server side rendering environments
+3. Frameworks with their own job scheduling system (like Ember)
+
+For example, a test environment might want to stub `IntersectionObserver`. Or in SSR mode, we might want to consider any element to be in the viewport for rendering purposes. Frameworks like Ember might want to ensure that any callback is scheduled via the framework's job scheduler.
+
+## Performance Optimizations
+
+There are two known performance optimizations that require this API change.
+
+### Using Native IntersectionObserver when available
+
+Whenever native `IntersectionObserver` is available, it should be used. There is an assumption that the native `IntersectionObserver` is more performant than the polyfill implementation. However, we don't want to ship the polyfill code to the browser when the browser already implements the native `IntersectionObserver`. We want to support apps that have per-browser builds.
+
+### The polling optimization
+
+The current `IntersectionObserver` polyfill checks every observed element every frame to see if the intersection state of an element has changed. Using requestAninmationFrame to poll on every frame requires significant CPU utilization. Meanwhile, it's not actually necessary to poll on every frame. If there were no changes to the window (scroll/resize) and no DOM changes, then there is no need to re-calculate intersection state.
+
+The solution should:
+- Fire all the events as expected/ as it was before.
+- Reduce CPU and memory utilization.
+- Should not create extra overhead while the system is idle.
+
+# Detailed design
+Rather than polling the viewport status of every time on every frame using `requestAnimationFrame`, Spaniel should just poll on the next animation frame after a scroll, resize or DOM change event occurs. To achieve this, we need to listen for scroll, resize and DOM change events, with respect to the viewport/[root](https://wicg.github.io/IntersectionObserver/#intersectionobserver-intersection-root). 
+
+We can listen for scroll or resize events using `addEventListener` function. But there is no universal way to listen for DOM changes, as [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) isn't universal. So we must provide an API through which the host app can provide a hook that allows Spaniel to listen for DOM changes.
```

Comment:
@stefanpenner I believe scroll and [resize](https://developer.mozilla.org/en-US/docs/Web/Events/resize) are [passive by default](https://developers.google.com/web/updates/2016/06/passive-event-listeners) since they are not cancellable events.

We are not recommending that people use `MutationObserver`. We can make it a future item to look into this. I believe we should recommend that the optimization only be used if you have a framework hook available like `Backburner.render`.

Reply:
> are passive by default since they are not cancellable events.

I wasn't aware that passive by default was cross browser. Have other browsers adopted this approach?

Reply:
> I believe we should recommend that the optimization only be used if you have a framework hook available like Backburner.render.

I'm not familiar with this API, were can I found out more about it? (We could add it, but I wasn't aware it existed)

Reply:
> were can I found out more about it?

@stefanpenner we used to use backburner to implement this back when `spaniel` used the same approach being re-introduced here. We used to monkeypatch `onBegin` but now I believe there's a [nice API for this](https://github.com/BackburnerJS/backburner.js/blob/cbff678057093ea88a2040bee8a9e21fc02742b9/lib/index.ts#L151).

> Have other browsers adopted this approach?

Not sure, couldn't find anything decisive. @krisselden do you know what browser compatibility is? @krisselden originally suggested that passive wasn't needed because it's default.

Reply:
> @stefanpenner we used to use backburner to implement this back when spaniel used the same approach being re-introduced here. We used to monkeypatch onBegin but now I believe there's a nice API for this.

This hook isn't for render, its just any BB flush, which may or may not include a render.